lib/Parser/Parse.cpp

```diff
@@ -2020,14 +2020,29 @@ void Parser::EnsureStackAvailable()
 
 void Parser::ThrowNewTargetSyntaxErrForGlobalScope()
 {
-    //TODO: (falotfi) we need reliably distinguish eval in global scope vs in a function
-    // The rule for this syntax error is any time new.target is called at global scope
-    // we are excluding new.target in eval at global scope for now.
-    if(GetCurrentNonLamdaFunctionNode() == nullptr && (this->m_grfscr & fscrEvalCode) == 0)
+    if (GetCurrentNonLamdaFunctionNode() != nullptr)
     {
-        Error(ERRInvalidNewTarget);
+        return;
     }
-}
+
+    if ((this->m_grfscr & fscrEvalCode) != 0 && (this->m_grfscr & fscrIndirect) == 0)
+    {
+        Js::JavascriptFunction * caller = nullptr;
+        if (Js::JavascriptStackWalker::GetNonLamdaCaller(&caller, m_scriptContext))
```

Comment:
Okay. That's not actually what we want. We want to find the JS caller, whatever it is, and if it's a lambda we want to look at the non-lambda function (if any) that encloses it lexically. So consider this case:

let l = () => { eval('new.target'); }
(function() { return l(); })();

And, conversely, this one:

var f = (function() { return () => { eval('new.target') } })();
f();